llpc/util/llpcElfWriter.cpp

```diff
@@ -735,6 +737,286 @@ template <class Elf> void ElfWriter<Elf>::updateMetaNote(Context *pContext, cons
+
+  // Clear the new section.
+  memset(newSection, 0, sizeof(SectionBuffer));
+
+  newSection->secHead = inputSection->secHead;
+
+  // Add the section name to .strtab section.
+  unsigned strTabSize = m_sections[m_strtabSecIdx].secHead.sh_size;
+  unsigned secNameSize = strlen(const_cast<char *>(sectionName)) + 1;
+
+  newSection->secHead.sh_name = strTabSize;
+  newSection->name = const_cast<char *>(sectionName);
+
+  newSection->secHead.sh_link = m_symSecIdx;
+  newSection->secHead.sh_size = 0;
+
+  auto strTabData = new uint8_t[strTabSize + secNameSize];
+  memcpy(strTabData, m_sections[m_strtabSecIdx].data, strTabSize);
+  memcpy(strTabData + strTabSize, sectionName, secNameSize);
+
+  delete[] m_sections[m_strtabSecIdx].data;
+
+  m_sections[m_strtabSecIdx].data = strTabData;
+  m_sections[m_strtabSecIdx].secHead.sh_size = strTabSize + secNameSize;
+}
+
+// =====================================================================================================================
+// Add the reloc related symbols and sections based on the input reloc
+//
+// @param reader : The ElfReader to the input fragment elf bin
+// @param inputReloc : The input reloc
+template <class Elf> void ElfWriter<Elf>::addRelocSymbols(const ElfReader<Elf> &reader, ElfReloc inputReloc) {
+  ElfSymbol relocSym = {};
+  reader.getSymbol(inputReloc.r_symbol, &relocSym);
+
+  // Create a section named origSectionName.cached.
+  char *secName = new char[strlen(relocSym.secName) + strlen(CachedRodataSectionSuffix) + 1];
+  strcpy(secName, relocSym.secName);
+  strcat(secName, CachedRodataSectionSuffix);
+
+  for (auto section : m_sections) {
+    if (strcmp(section.name, secName) == 0) {
+      delete[] secName;
+      return;
+    }
+  }
+
+  // If the section does not exist, create a new section and update the related symbols.
+  SectionBuffer newSection = {};
+  ElfSectionBuffer<Elf64::SectionHeader> *inputRodataSection = nullptr;
+  std::vector<ElfSymbol> inputRodataSymbols;
+  auto inputRodataSecIndex = reader.GetSectionIndex(relocSym.secName);
+  reader.getSectionDataBySectionIndex(inputRodataSecIndex, &inputRodataSection);
+  reader.GetSymbolsBySectionIndex(inputRodataSecIndex, inputRodataSymbols);
+
+  createNewSection(secName, inputRodataSection, &newSection);
+
+  // Update the data from the input section.
+  auto secData = new uint8_t[inputRodataSection->secHead.sh_size];
+  memcpy(secData, inputRodataSection->data, inputRodataSection->secHead.sh_size);
+  newSection.data = secData;
+  newSection.secHead.sh_size = inputRodataSection->secHead.sh_size;
+
+  m_sections.push_back(newSection);
+
+  // Update symbols.
+  for (auto rodataSymbol : inputRodataSymbols) {
+    ElfSymbol symbol = {};
+    // Create a new symbol named origSymbolName_cached.
+    char *newSymName = new char[strlen(rodataSymbol.pSymName) + strlen(CachedRodataSymbolSuffix) + 1];
+    strcpy(newSymName, rodataSymbol.pSymName);
+    strcat(newSymName, CachedRodataSymbolSuffix);
```

Comment:
Can you use `strncpy` and `strncat` to make sure that there's no buffer overflow here? Or even better, you could create a helper function that takes two c strings and returns a new array with the concatenated content?

Reply:
Also everywhere else. You can put this in llpcUtil or some other file and add unit tests.

Reply:
I guess using `StringRef` and `std::string` would be a lot nicer, but would require rewriting more of the `ElfWriter`. And that work is probably better spent on switching to the ElfLinker instead.

Reply:
As @Flakebi mentioned, I will use `strncpy` and `strncat` first.